src/main/groovy/io/openliberty/tools/gradle/tasks/GenerateFeaturesTask.groovy

```diff
@@ -301,4 +397,80 @@ class GenerateFeaturesTask extends AbstractFeatureTask {
     private getMPVersion(Object project) {
         return null;
     }
+
+    @SuppressWarnings("unchecked")
+    private Set<String> getConflicts(Throwable scannerResponse) {
+        try {
+            ClassLoader cl = getScannerClassLoader();
+            @SuppressWarnings("rawtypes")
+            Class featureConflictException = cl.loadClass("com.ibm.ws.report.exceptions.FeatureConflictException");
+            java.lang.reflect.Method conflictFeatureList = featureConflictException.getMethod("getFeatures");
+            if (conflictFeatureList == null) {
+                logger.debug("Error finding FeatureConflictException method getFeatures using reflection");
+                return null;
+            }
+            return (Set<String>) conflictFeatureList.invoke(scannerResponse);
+        } catch (ClassNotFoundException | MalformedURLException | NoSuchMethodException | SecurityException | IllegalAccessException | IllegalArgumentException | InvocationTargetException x) {
+            //TODO maybe nothing
+            logger.error("Exception:"+x.getClass().getName());
+            logger.error("Message:"+x.getMessage());
+            Object o = x.getCause();
+            if (o != null) {
+                logger.warn("Caused by exception:"+x.getCause().getClass().getName());
+                logger.warn("Caused by exception message:"+x.getCause().getMessage());
+            }
+        }
+        return null;
+    }
+
+    private Set<String> parseScannerMessage(String messages) {
+        Set<String> features = new HashSet<String>();
+        String[] messageArray = messages.split("\n");
+        for (String message : messageArray) {
+            if (message.startsWith("CWMIG12083")) {
+                String [] messageParts = message.split(" ");
+                if (messageParts.length > 4) { // should be 20
+                    features.add(messageParts[2]);
+                    features.add(messageParts[messageParts.length-2]);
+                }
+            }
+        }
+        return features;
+    }
+
+    // A class to pass the list of conflicts back to the caller.
+    private class NoRecommendationException extends Exception {
+        private static final long serialVersionUID = 1L;
+        Set<String> conflicts;
+        NoRecommendationException(Set<String> conflictSet) {
+            conflicts = conflictSet;
+        }
+        public Set<String> getConflicts() {
+            return conflicts;
+        }
+    }
+
+    // A class that encapsulates a list of conflicting features, a suggested list of replacements
+    // and a flag that indicates whether the conflicts were found in the features existing in the
+    // app's server config or if the conflicts exist in the binary files we examined.
+    private class RecommendationSetException extends Exception {
```

Comment:
We could also move these Exception classes to a util class in ci.common eventually

Reply:
Agreed.